src/librustdoc/doctree.rs

```diff
@@ -29,6 +59,63 @@ impl Module<'hir> {
             foreigns: Vec::new(),
             macros: Vec::new(),
             is_crate: false,
+            from_glob: false,
+        }
+    }
+
+    pub(crate) fn push_item(&mut self, ctx: &DocContext<'_>, new_item: Item<'hir>) {
+        let new_item_ns = ctx.tcx.def_kind(new_item.hir_item.def_id).ns();
+        if !new_item.name().is_empty() && new_item_ns.is_some() {
+            if let Some(existing_item) = self.items.iter_mut().find(|item| {
+                item.name() == new_item.name()
+                    && ctx.tcx.def_kind(item.hir_item.def_id).ns() == new_item_ns
+            }) {
+                match (existing_item.from_glob, new_item.from_glob) {
+                    (true, _) => {
+                        // `existing_item` is from glob, no matter whether `new_item` is from glob,
+                        // `new_item` should always shadow `existing_item`
+                        debug!("push_item: {:?} shadowed by {:?}", existing_item, new_item);
+                        *existing_item = new_item;
+                        return;
+                    }
+                    (false, true) => {
+                        // `existing_item` is not from glob but `new_item` is,
+                        // just keep `existing_item` and return at once
+                        return;
+                    }
+                    (false, false) => {
+                        // should report "defined multiple time" error before reach this
+                        unreachable!()
+                    }
+                }
+            }
+        }
+        // no item with same name and namespace exists, just collect `new_item`
+        self.items.push(new_item);
+    }
+
+    pub(crate) fn push_mod(&mut self, new_mod: Module<'hir>) {
+        if let Some(existing_mod) = self.mods.iter_mut().find(|mod_| mod_.name == new_mod.name) {
+            match (existing_mod.from_glob, new_mod.from_glob) {
+                (true, _) => {
+                    // `existing_mod` is from glob, no matter whether `new_mod` is from glob,
+                    // `new_mod` should always shadow `existing_mod`
+                    debug!("push_mod: {:?} shadowed by {:?}", existing_mod.name, new_mod.name);
+                    *existing_mod = new_mod;
```

Comment:
Hmm, if they're both from globs I'm not sure we should silently overwrite the two - maybe we should give an error or warning? When could this happen anyway, if you had `doc(inline)` on two different glob re-exports?

Reply:
An error will raise before reach here:
```
`xxx` is ambiguous (glob import vs glob import in the same module)
```
I'll make this an `unreachable!`